go/batch-submitter/drivers/sequencer/batch.go

```diff
@@ -0,0 +1,183 @@
+package sequencer
+
+import (
+	"errors"
+
+	l2types "github.com/ethereum-optimism/optimism/l2geth/core/types"
+)
+
+var (
+	// ErrBlockWithInvalidContext signals an attempt to generate a
+	// BatchContext that specifies a total of zero txs.
+	ErrBlockWithInvalidContext = errors.New("attempted to generate batch " +
+		"context with 0 queued and 0 sequenced txs")
+)
+
+// BatchElement reflects the contents of an atomic update to the L2 state.
+// Currently, each BatchElement is constructed from a single block containing
+// exactly one tx.
+type BatchElement struct {
+	// Timestamp is the L1 timestamp of the batch.
+	Timestamp uint64
+
+	// BlockNumber is the L1 BlockNumber of the batch.
+	BlockNumber uint64
+
+	// Tx is the optional transaction that was applied in this batch.
+	//
+	// NOTE: This field will only be populated for sequencer txs.
+	Tx *l2types.Transaction
+}
+
+// IsSequencerTx returns true if this batch contains a tx that needs to be
+// posted to the L1 CTC contract.
+func (b *BatchElement) IsSequencerTx() bool {
+	return b.Tx != nil
+}
+
+// BatchElementFromBlock constructs a BatchElement from a single L2 block. This
+// method expects that there is exactly ONE tx per block. The returned
+// BatchElement will reflect whether or not the lone tx is a sequencer tx or a
+// queued tx.
+func BatchElementFromBlock(block *l2types.Block) BatchElement {
+	tx := block.Transactions()[0]
```

Comment:
Do you think its a better idea to panic here in the off case that there is a block with no transactions? It should never happen but it seems to be safer just in case

Reply:
the logic as is will already panic if there are no transactions with index out of bounds. imo the more sinister case is if the block has _more_ than 1 transaction, as that would indicate we are silently skipping transactions. i'll add a panic to capture extra info in the empty case and also catch the case where are more than 1 transaction